R/get_tiles.R

```diff
@@ -0,0 +1,155 @@
+  # Get specified tiles
+  tiles_list <- list()
+  # TODO: add a progress bar
+  for (i in x) {
+    for (j in y) {
+      # Check if the tile exists
+      url <- paste0(
+        "https://tiles.macrostrat.org/", scale, "/",
+        zoom, "/", i, "/", j, ".mvt"
+      )
+      if (http_error(url)) {
+        stop("Tile does not exist: ", url)
+      }
+      tiles_list[[paste0(i, "/", j)]] <- read_mvt_sf(url)
+    }
+  }
+  # remove tiles without data
+  tiles_filt <- Filter(length, tiles_list)
+  # combine the tiles
+  tiles <- list()
+  # TODO: use st_union to combine duplicates based on map_id
+  lines <- lapply(tiles_filt, function(x) x$lines)
+  lines_filt <- Filter(Negate(is.null), lines)
+  tiles$lines <- st_rbindlist(lines_filt, use.names = TRUE, ignore.attr = TRUE,
+                              fill = TRUE, idcol = "tile")
+  units <- lapply(tiles_filt, function(x) x$units)
+  units_filt <- Filter(Negate(is.null), units)
+  tiles$units <- st_rbindlist(units_filt, use.names = TRUE, ignore.attr = TRUE,
+                              fill = TRUE, idcol = "tile")
+  return(tiles)
+  # TODO: document variables in return object
+}
+
+# MODIFIED FROM https://github.com/a-benini/sfhelpers
+#' @importFrom data.table rbindlist
+#' @importFrom sf st_as_sf st_crs st_geometry_type
+#' @importFrom uuid UUIDgenerate
+st_rbindlist <- function(l, ..., geometry_name = NULL) {
+  if (!is.list(l) || is.data.frame(l)) {
+    stop("input is ", class(l)[1L], " but should be a plain list of sf objects",
+         " to be stacked", call. = FALSE)
+  }
+  if (!(is.null(geometry_name) | (is.character(geometry_name) &
+                                 length(geometry_name) == 1))) {
+    stop("geometry_name must be either NULL or a single character string",
+         call. = FALSE)
+  }
+  is_not_null <- !vapply(l, is.null, logical(1))
+  if (!all(vapply(l[is_not_null], inherits, what = "sf", logical(1)))) {
+    stop("not all listed objects are of the class sf", call. = FALSE)
+  }
+  if(!any(is_not_null)) {
+    stop("no sf objects included in input list", call. = FALSE)
+  }
+  if (length(unique(lapply(l[is_not_null], st_crs))) > 1) {
+    stop("arguments have different crs", call. = FALSE)
+  }
+  # name of geometry column of the first listed sf-object
+  first_geometry_name <- attr(l[is_not_null][[1]], "sf_column")
+
+  # homogenize name of active geometry columns
+  if (length(unique(vapply(l[is_not_null], attr,
+                           which = "sf_column", character(1)))) > 1) {
+    tmp_geometry_name <- UUIDgenerate()
+    for (i in seq_along(l)[is_not_null]) {
+      st_geometry(l[[i]]) <- tmp_geometry_name
+    }
+  }
+
+  # homogenize geometry types
+  if (length(unique(vapply(l[is_not_null], st_geometry_type,
+                           by_geometry = FALSE, factor(1)))) > 1) {
+    for (i in seq_along(l)[is_not_null]) {
+      class(l[[i]][[attr(l[[i]], "sf_column")]])[[1]] <- "sfc_GEOMETRY"
+    }
+  }
+
+  sf <- st_as_sf(rbindlist(l, ...))
```

Comment:
Is the value of using `rbindlist` over `do.call(rind,list)` efficiency? Again, just thinking about reducing imports. If `rbindlist` is much faster I guess it makes sense given the size of data a user could request.

Reply:
And here's zoom = 3:
![image](https://github.com/user-attachments/assets/12afbee4-d9de-480a-9e9e-2f6031edad7d)